modules/issue_tracker/php/endpoints/attachment.class.inc

```diff
@@ -0,0 +1,245 @@
+        case 'GET':
+            return $this->_handleGET($request);
+        case 'POST':
+            return $this->_handlePOST($request);
+        case 'DELETE':
+            return $this->_handleDELETE($request);
+        default:
+            return new \LORIS\Http\Response\JSON\MethodNotAllowed(
+                $this->allowedMethods()
+            );
+        }
+    }
+
+    /**
+     * Handle GET requests for attachment
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handleGET(ServerRequestInterface $request) : ResponseInterface
+    {
+        // Parse GET query params.
+        $values = $request->getQueryParams();
+
+        $file_hash = $values['file_hash'];
+        $filename  = $values['filename'];
+        $mime_type = $values['mime_type'];
+
+        $config = \NDB_Config::singleton();
+        $attachment_data_dir = rtrim(
+            $config->getSetting('IssueTrackerDataPath'),
+            '/'
+        );
+
+        $fileToDownload = trim($attachment_data_dir) .
+            '/attachments/' .
+            $file_hash;
+
+        if (!is_readable($fileToDownload)) {
+            throw new \LorisException('Forbidden Access');
+        }
+
+        $name = rawurldecode($filename);
+
+        return (new \LORIS\Http\Response())
+            ->withHeader('Content-Type', $mime_type)
+            ->withHeader(
+                'Content-Disposition',
+                'attachment; filename=' . basename($name)
+            )
+            ->withStatus(200)
+            ->withBody(new \LORIS\Http\FileStream($fileToDownload));
+    }
+
+    /**
+     * Handle POST requests for attachment
+     * and used for uploading attachment file.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handlePOST(ServerRequestInterface $request) : ResponseInterface
+    {
+        // Parse POST request body.
+        $values     = $request->getQueryParams();
+        $user       = \NDB_Factory::singleton()->user();
+        $attachment = new \LORIS\issue_tracker\UploadHelper();
+        $response   = $attachment->setupUploading(
+            $user,
+            $_FILES,
+            $values
+        );
+        return new \LORIS\Http\Response\JsonResponse(
+            $response
+        );
+    }
+
+    /**
+     * Handle DELETE requests for attachment
+     * and marks attachment as deleted from issue_tracker.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    private function _handleDELETE(ServerRequestInterface $request)
+    : ResponseInterface
+    {
+        // Verify User can delete file.
+        if ($this->_deletePermission($request)) {
```

Comment:
I would suggest doing something like:
```
if ($this->attachment->getCreator() != $request->getAttribute('user')) {
    return new \LORIS\Http\Response\JSON\Forbidden();
}
try {
   $this->attachement->delete();
} catch (...) {
    return new \LORIS\Http\Response\JSON\...someErrorResponse();
}

return new \LORIS\Http\Response\JSON\NoContent();
```
That would mean that `LORIS\issue_tracker\endpoints\Attachment` would have a constructor that takes a `LORIS\issue_tracker\Models\Attachment` as a parameter. I know it does not exists but it would also be a good placeholder for getETag, getFileMimeType and your CRUD functions to reside. 

`LORIS\issue_tracker\Module` would need to instantiate the Attachment model if ID is provided in the queryParams

The reason is that we try to avoid SQL queries in Endpoints.

Reply:
This is much more clear to me.

Reply:
@johnsaigle @xlecours that breaks it. I don't find my code that hard to understand and it works. 😐

Reply:
resolved offline